dotnet/Cucumber.Messages/generated/TestStep.cs

```diff
@@ -49,11 +49,11 @@ List<StepMatchArgumentsList> stepMatchArgumentsLists
     ) 
     {
               this.HookId = hookId;
-              RequireNonNull<string>(id, "Id", "TestStep.Id cannot be null");
+        RequireNonNull<string>(id, "Id", "TestStep.Id cannot be null");
         this.Id = id;
               this.PickleStepId = pickleStepId;
-              this.StepDefinitionIds = new List<string>(stepDefinitionIds);
-              this.StepMatchArgumentsLists = new List<StepMatchArgumentsList>(stepMatchArgumentsLists);
+              this.StepDefinitionIds = stepDefinitionIds == null ? new List<string>() : new List<string>(stepDefinitionIds);
```

Comment:
I think this should be:

```suggestion
              this.StepDefinitionIds = stepDefinitionIds == null ? null : new List<string>(stepDefinitionIds);
```

Similar to https://github.com/cucumber/messages/blob/de3f88f1b7fa43ebceeb6e247767e44ff8c18730/java/src/generated/java/io/cucumber/messages/types/TestStep.java#L36C1-L36C122

The field is not required, so it can be null. When serializing, if configured correctly, this will result in the omission of the field.

Reply:
In principle, I agree with you, purely optional elements should be null and not emitted when serialized. 
I'm curious then about these lines in the Hooks example in the CCK:
```
{"testCase":{"id":"33","pickleId":"20","testSteps":[{"hookId":"0","id":"29"},{"hookId":"1","id":"30"},{"id":"31","pickleStepId":"19","stepDefinitionIds":["2"],"stepMatchArgumentsLists":[{"stepMatchArguments":[]}]},{"hookId":"4","id":"32"}]}}
{"testCase":{"id":"38","pickleId":"22","testSteps":[{"hookId":"0","id":"34"},{"hookId":"1","id":"35"},{"id":"36","pickleStepId":"21","stepDefinitionIds":["3"],"stepMatchArgumentsLists":[{"stepMatchArguments":[]}]},{"hookId":"4","id":"37"}]}}
{"testCase":{"id":"43","pickleId":"24","testSteps":[{"hookId":"0","id":"39"},{"hookId":"1","id":"40"},{"id":"41","pickleStepId":"23","stepDefinitionIds":[],"stepMatchArgumentsLists":[]},{"hookId":"4","id":"42"}]}}
``` 
In the first two lines, the third step of each test case is matched with a StepDefinition; each of those stepDefinitions accept no parameters. So why are the stepMatchArgumentsLists and stepMatchArguments properties being serialized as empty lists? Why not omit them altogether? (They are optional according to the schema).
Likewise in the third example (the test case does not bind to any step definition (aka, pending). Why is the StepDefinitionIds array getting serialized when it is empty?

I will try out your proposal and see how well it works with Reqnroll.

Reply:
A `TestStep` is either a `PickleTestStep` or a `HookTestStep`.

 * When a `PickleTestStep` then the  `id`, `pickleStepId`, `stepDefinitionIds` and `stepMatchArgumentsLists` are required. 
 * When a `HookTestStep` then the `id` and `hookId` are required.

So in the union of these, only the `id` is required.

There are now better ways to express this in a json schemas but either we didn't have those at the time, or they didn't quite work.

Reply:
Oof. That makes a general solution almost impossible. 
If the template were to deserialize missing optional lists as null then they won't round-trip correctly. (Nor would it round-trip correctly with the code as it exists or with what I had proposed)
I'll consider inserting a patch up phase in the Serializer helper class.

Reply:
How does it make round tripping correctly impossible? Not including null values when serializing and not replacing absent/null values with empty list during deserialisation is a successful round trip.

 Or do you mean with the current implementation?

Reply:
There might be a use-case for it, but it would be somewhat out of scope for this PR. 

In this example, an error correction to an empty list would also be context sensitive. For example the correct solution would depend on whether we are dealing with a `PickleTestStep` or a `HookTestStep`.